### PR TITLE
[TASK] rewrite forcegraph with d3 version 4 

### PR DESCRIPTION
**Pull request by genofire**
## Description
Update d3 to version v4

## Motivation and Context
Better performance with modular d3

## How Has This Been Tested?
Not, still in progress.

## Checklist:

- [x] My code follows the code style of this project. (CI will test it anyway and also needs approval)
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

## TODO to new issues
- [x] extract color const in #91
- [x] vpn link boolean (instatt `d.o.type === 'fastd' || d.o.type === 'L2TP'`) in #92 

## TODO for this PR
- [x] css canvas width (xf-: you need width and height inline, style is maybe not necessary)
- [x] delete comment select vpn link
- [x] Force of meshes - large local clouds are far away
- [x] Start position http://i.imgur.com/70RinW4.png

### CODE REVIEW

**Comment by genofire** [comment]
comment for myself :)
build a own d3 main required definition ...
```
var d3 = Object.assign({}, require("d3-format"), require("d3-geo"), require("d3-geo-projection"));
```

https://github.com/d3/d3/blob/b2456e0508a325ea8e9945070b86e7ed7c83ffc2/README.md

No access to the events :(

File: lib/forcegraph/draw.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+define(['helper'], function (helper) {
+  var self = this;
+
+  var ctx;
+  var transform;
+
+  var highlight;
+
+  const nonUplinkColor = '#F2E3C6';
```
Comment (xf-) [inline lib/forcegraph/draw.js]: lowercase hex - maybe we find a way to move all hardcoded canvas stuff into a setting js.

File: lib/forcegraph/draw.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  const locationColor = '#00ff00';
+  const noLocationColor = '#0000ff';
+  const clientColor = 'rgba(230, 50, 75, 1.0)';
```
Comment (xf-) [inline lib/forcegraph/draw.js]: opacity 1 - please use hex

File: lib/forcegraph/draw.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+  const cableColor = '#50B0F0';
```
Comment (xf-) [inline lib/forcegraph/draw.js]: Maybe we should think about colors or styles (line style ...)

File: lib/forcegraph/draw.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      helper.positionClients(ctx, d, Math.PI, d.o.node.statistics.clients, 15);
+      ctx.fillStyle = clientColor;
+      ctx.fill();
+      ctx.beginPath();
+      var name = d.o.node_id;
+      if (d.o.node && d.o.node.nodeinfo) {
+        name = d.o.node.nodeinfo.hostname;
+      }
+      ctx.textAlign = 'center';
+      ctx.fillStyle = '#fff';
```
Comment (xf-) [inline lib/forcegraph/draw.js]: duplicate - set in line 26 - colors should be const or config

File: lib/forcegraph/draw.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  self.drawLink = function drawLink(d) {
+    ctx.beginPath();
+    ctx.moveTo(d.source.x, d.source.y);
+    var to = [d.target.x, d.target.y];
+
+    to = drawHighlightLink(d, to);
+
+    ctx.lineTo(to[0], to[1]);
+    ctx.strokeStyle = d.o.type === 'Kabel' ? cableColor : d.color;
+    if (d.o.type === 'fastd' || d.o.type === 'L2TP') {
```
Comment (xf-) [inline lib/forcegraph/draw.js:92]: We use this condition multiple times - maybe create a helper

File: lib/forcegraph/math.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    /* http://stackoverflow.com/questions/849211 */
+    var l2 = distance(a, b);
+    if (l2 === 0) {
+      return distance(p, a);
+    }
+    var t = ((p.x - a.x) * (b.x - a.x) + (p.y - a.y) * (b.y - a.y)) / l2;
+    if (t < 0) {
+      return distance(p, a);
+    }
+    if (t > 1) {
```
Comment (xf-) [inline lib/forcegraph/math.js]: else if (will not be validated if first if matches)

File: lib/forcegraph/draw.js [hunk 7]
```diff
@@ -0,0 +1,109 @@
+'use strict';
```
Comment (xf-) [inline lib/forcegraph/draw.js]: all over use strict inside define (can't remember why)

File: lib/forcegraph.js [hunk 8]
```diff
@@ -1,744 +1,200 @@
-define(['d3', 'helper'], function (d3, helper) {
+define(['d3-selection', 'd3-force', 'd3-zoom', 'd3-drag', 'forcegraph/math', 'forcegraph/draw'], function (d3Selection, d3Force, d3Zoom, d3Drag, math, draw) {
```
Comment (xf-) [inline lib/forcegraph.js]: 2 kines, avoid scrollbars

File: lib/forcegraph.js [hunk 9]
```diff
@@ -1,744 +1,200 @@
...
   'use strict';
 
-  var margin = 200;
-  var NODE_RADIUS = 15;
-  var LINE_RADIUS = 12;
-
   return function (config, linkScale, sidebar, router) {
     var self = this;
+    var el;
```
Comment (xf-) [inline lib/forcegraph.js]: If we use `const`, we should consider `let`
https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Statements/let

File: lib/forcegraph.js [hunk 10]
```diff
@@ -1,744 +1,200 @@
...
 
-    function nodeName(d) {
-      if (d.o.node && d.o.node.nodeinfo) {
-        return d.o.node.nodeinfo.hostname;
-      }
-      return d.o.id;
+    function resizeCanvas() {
+      canvas.width = el.offsetWidth;
+      canvas.height = el.offsetHeight;
+      canvas.style.width = el.offsetWidth + 'px';
```
Comment (xf-) [inline lib/forcegraph.js]: style is unnecessary - 100vw 100vh if we need it. (height, width attribute is necessary)

File: lib/forcegraph.js [hunk 11]
```diff
@@ -1,744 +1,226 @@
...
-    }
+    draw.setTransform(transform);
 
     function resizeCanvas() {
-      var r = window.devicePixelRatio;
-      canvas.width = el.offsetWidth * r;
-      canvas.height = el.offsetHeight * r;
+      canvas.width = el.offsetWidth;
+      canvas.height = el.offsetHeight;
       canvas.style.width = el.offsetWidth + 'px';
```
Comment (xf-) [inline lib/forcegraph.js]: not needed, could be done via CSS

File: lib/forcegraph.js [hunk 12]
```diff
@@ -1,744 +1,226 @@
...
-
-      var sidebarWidth = sidebar();
-      var delta = [size[0] - newSize[0], size[1] - newSize[1]];
-      var translate = z.translate();
-      var translateNew = [sidebarWidth + (translate[0] - sidebarWidth - delta[0] / 2) * scale, (translate[1] - delta[1] / 2) * scale];
+      var closedLink;
+      var radius = LINK_RADIUS_SELECT;
+      intLinks
+        /* Disable Clickable VPN
+        .filter(function (d) {
```
Comment (xf-) [inline lib/forcegraph.js]: Remove unnecessary code

File: lib/forcegraph.js [hunk 13]
```diff
@@ -1,744 +1,226 @@
...
 
-    font = window.getComputedStyle(el).fontSize + ' ' + window.getComputedStyle(el).fontFamily;
-
-    zoomBehavior = d3.behavior.zoom()
-      .scaleExtent([1 / 3, 3])
-      .on('zoom', onPanZoom)
-      .translate([sidebar(), 0]);
+    forceLink = d3Force.forceLink()
+     .distance(function (d) {
+       if (d.o.type === 'fastd' || d.o.type === 'L2TP') {
```
Comment (xf-) [inline lib/forcegraph.js]: Should be moved to helpers or similar or do we have a field Boolean vpn?

File: lib/forcegraph.js [hunk 14]
```diff
@@ -1,744 +1,226 @@
...
+     })
+     .strength(function (d) {
+       if (d.o.type === 'fastd' || d.o.type === 'L2TP') {
+         return 0.02;
+       }
+       return Math.max(0.5, 1 / d.o.tq);
+     });
+
+    var zoom = d3Zoom.zoom()
+         .scaleExtent([ZOOM_MIN, ZOOM_MAX])
```
Comment (xf-) [inline lib/forcegraph.js]: to many indents

File: lib/forcegraph.js [hunk 15]
```diff
@@ -1,744 +1,226 @@
...
+           draw.setTransform(transform);
+           redraw();
+         });
+
+
+    force = d3Force.forceSimulation()
+      .force('link', forceLink)
+      .force('charge', d3Force.forceManyBody())
+      .on('tick', redraw)
+      .alphaTarget(0.1);
```
Comment (xf-) [inline lib/forcegraph.js]: const, avoid hardcoded settings in middle of code

File: lib/forcegraph.js [hunk 16]
```diff
@@ -1,744 +1,226 @@
...
+
+    var drag = d3Drag.drag()
+      .subject(function () {
```
Comment (xf-) [inline lib/forcegraph.js]: indents

**Review by xf- — CHANGES_REQUESTED** [review]

File: lib/forcegraph/draw.js [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+    }
+    return to;
+  }
+
+  self.drawNode = function drawNode(d) {
+    ctx.beginPath();
+
+    drawHighlightNode(d);
+
+    ctx.moveTo(d.x + 3, d.y);
```
Comment (xf-) [inline lib/forcegraph/draw.js:69]: wie findest?

```javascript
    ctx.moveTo(d.x + 3, d.y);
    ctx.arc(d.x, d.y, 6, 0, 2 * Math.PI);
    ctx.fillStyle = nonUplinkColor;
    ctx.globalAlpha = 1;
    ctx.fill();
```

File: lib/forcegraph.js (hunk 12 above)
Comment (genofire) [reply lib/forcegraph.js]: Should we allow to select vpn links? It is not in the develop branch.

File: lib/forcegraph/draw.js (hunk 3 above)
Comment (genofire) [reply lib/forcegraph/draw.js]: Jep, also for https://github.com/ffrgb/meshviewer/blob/develop/lib/map/clientlayer.js#L46 -> later

File: lib/forcegraph.js (hunk 13 above)
Comment (genofire) [reply lib/forcegraph.js]: boolean :+1:

File: lib/forcegraph.js (hunk 16 above)
Comment (genofire) [reply lib/forcegraph.js]: hier nicht korrekt?

File: lib/forcegraph/draw.js (hunk 17 above)
Comment (genofire) [reply lib/forcegraph/draw.js:69]: done

File: lib/forcegraph.js (hunk 9 above)
Comment (genofire) [reply lib/forcegraph.js]: back to `var` is it okay?

Comment (xf-) [reply lib/forcegraph.js]: i prefer let and const, but if we rewrite complete file, we should be consistent.

File: lib/forcegraph.js (hunk 12 above)
Comment (xf-) [reply lib/forcegraph.js]: why not - vpn has also quality

File: lib/forcegraph.js (hunk 16 above)
Comment (xf-) [reply lib/forcegraph.js]: allgemein passt das nicht, glaube function is nur ein space, paar doppelte leerzeilen

File: lib/forcegraph/math.js [hunk 18]
```diff
@@ -0,0 +1,31 @@
+define(function () {
+  let self = this;
```
Comment (xf-) [inline lib/forcegraph/math.js]: let won't work :( - requirejs with uglify trhows an error

File: lib/forcegraph/draw.js [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+    ctx.arc(d.x, d.y, 8, 0, 2 * Math.PI);
+    ctx.fillStyle = nodeColor;
+    ctx.fill();
+
+    drawDetailNode(d);
+  };
+
+  self.drawLink = function drawLink(d) {
+    if (d.source.x < -width && d.target.x < -width || d.source.y < -height && d.target.y < -height ||
```
Comment (xf-) [inline lib/forcegraph/draw.js]: das muesste bei jedem drag ausgefuehrt werden, sonst is ein teil der nodes eventuell nie sichtbar. weis nicht ob das soviel bringt.

Comment (genofire) [reply lib/forcegraph/draw.js]: Wird es, durch den `.tick` im `forcegraph.js`.
Doch ich hab nicht auf transform (z.B. Zoom) geachtet.

Comment (genofire) [reply lib/forcegraph/draw.js]: Ähnlich gab es dies auch bereits vorher mit ein filter und der `function visibleNodes(d) {` - doch wenn jeweils erst ein filter und dann ein foreach kommt, ist dies nicht wirklich performant

File: lib/forcegraph/draw.js (hunk 5 above)
Comment (genofire) [reply lib/forcegraph/draw.js:92]: Will make a PR (Issue) for other components, too.
I am for a boolean variable.

File: lib/forcegraph/draw.js [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+      ctx.globalAlpha = 0.8;
+      ctx.lineWidth = 2.5;
+    }
+    ctx.stroke();
+  };
+
+  self.setCTX = function setCTX(newValue) {
+    ctx = newValue;
+  };
+  self.setHighlight = function setHighlight(newValue) {
```
Comment (xf-) [inline lib/forcegraph/draw.js]: Add a empty line between functions

**Comment by xf-** [comment]
merged via 98f175d

**Pull request closed without merging**